apm-agent-core/src/main/java/co/elastic/apm/agent/bci/bytebuddy/InstallationListenerImpl.java

```diff
@@ -0,0 +1,79 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package co.elastic.apm.agent.bci.bytebuddy;
+
+import net.bytebuddy.agent.builder.AgentBuilder;
+import net.bytebuddy.agent.builder.ResettableClassFileTransformer;
+import net.bytebuddy.dynamic.loading.ByteArrayClassLoader;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.lang.reflect.Method;
+import java.util.HashMap;
+import java.util.Map;
+
+public class InstallationListenerImpl extends AgentBuilder.InstallationListener.Adapter {
+
+    @Override
+    public void onAfterWarmUp(Map<Class<?>, byte[]> types, ResettableClassFileTransformer classFileTransformer, boolean transformed) {
+        Logger logger = null;
+        try {
+            // should be safe to get a logger by now
+            logger = LoggerFactory.getLogger(InstallationListenerImpl.class);
+            if (!transformed) {
+                logger.warn("Byte Buddy warmup ended without transforming at least one class. The agent may not work as expected.");
+            } else  {
+                byte[] transformedInstrumentedClass = null;
+                for (Map.Entry<Class<?>, byte[]> classEntry : types.entrySet()) {
+                    if (classEntry.getKey().equals(Instrumented.class)) {
+                        transformedInstrumentedClass = classEntry.getValue();
+                        break;
+                    }
+                }
+
+                String instrumentedClassName = Instrumented.class.getName();
+                if (transformedInstrumentedClass != null) {
+                    logger.debug("Bytecode transformation {} succeeded", instrumentedClassName);
+
+                    // warm up the complete invokedynamic linkage route
+                    HashMap<String, byte[]> typeDefinitions = new HashMap<>();
+                    typeDefinitions.put(instrumentedClassName, transformedInstrumentedClass);
+                    ByteArrayClassLoader.ChildFirst childFirstCL = new ByteArrayClassLoader.ChildFirst(null, typeDefinitions);
```

Comment:
Why can't we just call `Instrumented.isInstrumented()`? Why do we need to create a dedicated class loader?

Reply:
Warmup doesn't retransform the class, it only applies the bytecode transformation. 
In order to retransform classes, it would have to register the BB class file transformer, thus kicking the instrumentation race when invoking `retransformClasses`, which misses the point of avoiding deadlocks.
This allows me to warmup the entire `invokedynamic` linkage path on the attaching thread before installing the BB agent.

Reply:
Ah, understood. A small comment about that in the code would be nice.